Add py-tools variant and set $ISSM_DIR Environment Variable

lawrenceabird opened the pull request:
This PR adds the following functionality:
- New `py-tools` variant to install Python function files to `<prexix>/python-tools/`. These files are necessary to A) access the wrappers that are included in `<prefix>/lib/` and B) to conduct various pre- and post-processing of ISSM models. Note: Python files are stored in various subdirectories (e.g. `src/m/mesh`, `src/m/mask`, `src/m/solve`) in the ISSM directory. Here, we flatten this structure to just save all Python files in one directory (`<prefix>/python-tools/`) as the subdirectory structure serves no functional purpose.

- Export `$ISSM_DIR` environment variable. This is required by various Python wrappers to obtain information about the run environment.

This PR addresses Issue #333.

Functionality was tested in https://github.com/ACCESS-NRI/ACCESS-ISSM/pull/25

aidanheerdegen commented on packages/issm/package.py:
```diff
@@ -198,7 +212,39 @@ def configure_args(self):
+                for file in files:
+                    if file.endswith(".py"):
+                        src_file = join_path(root, file)
+                        install(src_file, py_dst)
```
Given this is 528 files, which could end up using a fair bit of inode quota if there are many pre-release builds in a PR, could you [bundle them all into a zip file](https://realpython.com/python-zip-import/) and add that to your `PYTHONPATH`?

It is [automatically supported by python](https://docs.python.org/3/library/zipimport.html) (since v2.3 apparently!).

lawrenceabird replied:
Interesting! I didn't know this. I can have a play with this if we think it's necessary. Or, I can just clean up the relic pre-lease builds in the PR...?

lawrenceabird replied:
Sounds good! I'll have a crack.

aidanheerdegen replied:
Great, thanks.

lawrenceabird commented on packages/issm/package.py:
```diff
@@ -94,10 +101,17 @@ class Issm(AutotoolsPackage):
     conflicts("%gcc@14:", msg="ISSM cannot be built with GCC versions above 13")
 
+    # +wrappers requires +py-tools to access the wrappers
+    conflicts("+wrappers", when="~py-tools", msg="The +wrappers variant requires +py-tools")
```
Yea, I'll add this. It isn't strictly *required*, but it sure makes the Python tools more helpful is you can access the wrappers too.

lawrenceabird commented:
This new functionality was used in [this PR](https://github.com/ACCESS-NRI/ACCESS-ISSM/pull/26) to test a new deployment of an updated ISSM code base on Gadi. This deployed successfully.

aidanheerdegen reviewed (dismissed):
Looks good.

harshula commented on packages/issm/package.py:
```diff
@@ -198,7 +216,44 @@ def configure_args(self):
+
+            # Create a zip file and add *.py files, excluding exclude_dirs
+            with zipfile.ZipFile(py_dst, 'w', zipfile.ZIP_DEFLATED) as zf:
+                for root, dirs, files in os.walk(py_src):
```
Have you considered using Spack's `find()` function?

lawrenceabird commented:
@harshula and @aidanheerdegen -- I've addressed all comments and confirmed that these changes still work ([here](https://github.com/ACCESS-NRI/ACCESS-ISSM/pull/25)). I subsequently added an `access-release` version in preparation for a new "release" branch in our ISSM forked repo. This untested addition should have no impact on this PR.

harshula commented on packages/issm/package.py:
```diff
@@ -198,7 +216,51 @@ def configure_args(self):
+            py_files = [
+                f for f in py_files
+                if not any(f"/{excl}/" in f for excl in exclude_dirs)
+            ]
```
Hi @lawrenceabird , Yes, the additional filesystem activity is not significant. @becker33 provided some suggested code that is going to be tried next.

aidanheerdegen replied:
I'm going to resolve this as @lawrenceabird has addressed the review and would like to get this merged. We can revisit if necessary.
